src/nnet3bin/nnet3-xvector-get-egs-seg.cc

```diff
@@ -0,0 +1,167 @@
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include <sstream>
+#include "util/common-utils.h"
+#include "nnet3/nnet-example.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+// Process the utt2label file and store it as a map from utt_id to
+// label
+static void ProcessUttToLabelFile(const std::string &utt2label_rxfilename, 
+    unordered_map<std::string, int32, StringHasher> *utt_to_label) {
+  Input utt2label_input(utt2label_rxfilename);
+  if (!utt2label_rxfilename.empty()) {
+    std::string line;
+    while (std::getline(utt2label_input.Stream(), line)) {
+      std::vector<std::string> fields;
+      SplitStringToVector(line, " \t\n\r", true, &fields);
+      if (fields.size() != 2) {
+        KALDI_ERR << "Expected 2 fields in line of utt2label file, got "
+                  << fields.size() << " instead.";
+      }
+      std::string utt_id = fields[0];
+      int32 label_id;
+      if (!ConvertStringToInteger(fields[1], &label_id)) {
+        KALDI_ERR << "Expect integer for label_id";
+      }
+      // Add to map
+      (*utt_to_label)[utt_id] = label_id;
+    } // end of while 
+  } // end of if
+}
+
+
+static void WriteExample(const MatrixBase<BaseFloat> &feats,
+  const int32 this_pdf_id, const std::string &key,
+  bool compress, int32 num_pdfs, int32 &num_egs_written,
+  NnetExampleWriter *example_writer) {
+
+  NnetIo nnet_input = NnetIo("input", 0, feats);
+  for (std::vector<Index>::iterator indx_it = nnet_input.indexes.begin();
+      indx_it != nnet_input.indexes.end(); ++indx_it) {
+    indx_it->n = 0;
+  }
+  Posterior label;
+  std::vector<std::pair<int32, BaseFloat> > post;
+  post.push_back(std::pair<int32, BaseFloat>(this_pdf_id, 1.0));
+  label.push_back(post);
+  NnetExample eg;
+  eg.io.push_back(nnet_input);
+  eg.io.push_back(NnetIo("output", num_pdfs, 0, label));
+  if (compress) {
+    eg.Compress();
+  }
+
+  example_writer->Write(key, eg);
+  num_egs_written++;
+}
+
+
+} // namespace nnet3
+} // namespace kaldi
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace kaldi::nnet3;
+    typedef kaldi::int32 int32;
+
+    const char *usage =
+        "In previous xvector setup, the different examples in the same archive\n"
+        "are equvalent length. Now, we hope the length of different examples in\n"
+        "the same archive is different, so that the randomness of data is increased\n"
+        "For this binary, it deals with the variable length feats.scp file.\n"
+        "Generate the variable length egs into egs.ark file.\n"
+        "Usage: nnet3-xvector-get-egs [options] <utt2label-rxfilename>"
```

Comment:
you should also describe that in old setup uses ranges.* files to generate archives, which contains local and global archive index and example line in each line. In new setup is similar to conventional nnet3-get-egs setup and it uses separate input and output files to generate examples.